### DISCUSSION TITLE
How to programatically create and copy route URLs?

### DISCUSSION THREAD

**Discussion by jameslawson** [TanStack/router#1391] [Q&A]
### Context

I'm migrating my app from react-router to tanstack/router. I'm current stuck migrating a feature where users can click a menu item that copies a URL to the clipboard. 
<img width="216" alt="image" src="https://github.com/TanStack/router/assets/1913533/d2523416-24f9-4f59-922f-23bbd4a7d061">

This menu component is reusable and used in a few screens: the component itself doesn't know exactly what to copy, this info passed down via a React prop.

With react-router:
-  a parent component makes a call to [`generatePath`](https://v5.reactrouter.com/web/api/generatePath) to generate a string
- this was passed down to the component. The component itself accepts this as a prop `url`. 
- when the appropriate menu item is selected ("Copy Permalink") it copies this string to the clipboard. 
- I didn't simply copy `window.location.href` to clipboard because for this particular app, only certain parts of the URL are important for the permalink and other parts should be ignored

**Question**: What's the best approach for migrating this to tanstack-router?

----

### Code

**Before migration (react-router)**
```typescript
interface Props {
  url: string;
  // ....
}
export default function MoreMenu({ url }: Props) { 
   // ...
  const onCopyClick = () => {
       copyToClipboard(url);
   }
}  
```

**After migration (tanstack/router) ??**

I figured I could pass ToOptions down to the component and let it generate the full URL. But I'm encountering type errors.

```typescript
import type { ToOptions } from '#/app/router';
import { useLinkProps } from '@tanstack/react-router';

interface Props {
  toOpts: ToOptions;
  // ....
}
export default function MoreMenu({
   toOpts
   // ...
}: Props) {
   // ...
   const props = useLinkProps(toOpts);
                           // ^~~~~~ ERROR here [*]
   const href = props.href;
   const onCopyClick = () => {
       copyToClipboard(href);
   }
    // ...
}
```
<details>
<summary>app/router.ts</summary>

```typescript
import type {
  RouteIds as RouteIdsTanStack,
  ToOptions as ToOptionsTanStack,
} from '@tanstack/react-router';
import { createRouter } from '@tanstack/react-router';
import { rootRoute } from './router/index';

// #/app/router/router.ts
const routeTree = rootRoute.addChildren([
  // ...
]);
const router = createRouter({ routeTree });

export { router };

export type RouteTree = typeof routeTree;
export type RouteIds = RouteIdsTanStack<typeof routeTree>;
export type ToOptions = ToOptionsTanStack<RouteTree>;
```
</details>

It seems I can't pass an object of type `ToOptions` to `useLinkProps` (which is little bit unexpected for me)?

```
Argument of type 'ToOptions' is not assignable to parameter of type 'UseLinkPropsOptions<Route<any, "/", "/", string, "__root__", RootSearchSchema, RootSearchSchema, RootSearchSchema, RootSearchSchema, ... 11 more ..., any>, string, "" | ... 29 more ... | "/privacy", string, "">'.
  Type 'ToOptions' is not assignable to type 'CheckPathError<Route<any, "/", "/", string, "__root__", RootSearchSchema, RootSearchSchema, RootSearchSchema, RootSearchSchema, ... 11 more ..., any>>'.
    Types of property 'to' are incompatible.
      Type 'ToPathOption<Route<any, "/", "/", string, "__root__", RootSearchSchema, RootSearchSchema, RootSearchSchema, RootSearchSchema, ... 11 more ..., any>, string, ""> | undefined' is not assignable to type 'RoutePaths<Route<any, "/", "/", string, "__root__", RootSearchSchema, RootSearchSchema, RootSearchSchema, RootSearchSchema, ... 11 more ..., any>>'.
        Type 'undefined' is not assignable to type 'RoutePaths<Route<any, "/", "/", string, "__root__", RootSearchSchema, RootSearchSchema, RootSearchSchema, RootSearchSchema, ... 11 more ..., any>>'
```

I can't find any examples for the intended use of `useLinkProps`, are we supposed to be able to pass `ToOptions` to `useLinkProps`? And if not is there an alternative approach where I can programatically generate a full url using the state: from/to/params/search? The ancestor react components have access to all this state data and can pass down what's needed. Here I'm looking for the simplest and most idiomatic way for this component to programatically generating the URL given this state with tanstack/router.

Loving tanstack/router so far! Thank you for your time! 🙏 

Versions:
```
"@tanstack/react-router": "1.24.0"
"@tanstack/router": "0.0.1-beta.53"
"react" "18.2.0"
"typescript": "5.3.3"
Google Chrome Version: 123.0.6312.58
 ```

**Comment by jedwards1211** [comment]
@tannerlinsley for testing we need a static way to build a URL for Playwright or Selenium to navigate to from a TanStack route and parameters.  We can't rely on `.Link` or anything that only works within a React component to build these URLs.

**Reply by schiller-manuel** [reply]
router.buildLocation() does this

**Reply by jedwards1211** [reply]
I guess as long as we can import the router object we created in our testing code we can build the URL?  It seems like not much consideration has been given to testing needs yet

**Reply by schiller-manuel** [reply]
if you want to build a link programatically, you need the router instance.

**Reply by jedwards1211** [reply]
@schiller-manuel Ideally testing code would be able to import the router module without depending on a transpiler to transpile TSX, but the examples I see have TSX in all of the file routes.  Maybe exporting `React.lazy(() => ...)` from the file routes would probably work, but I'm not sure yet how the codegen identifies route components?

**Reply by schiller-manuel** [reply]
vitest just runs TSX fine in tests, what are you using?

**Reply by jedwards1211** [reply]
I have always used different tools like React Router, Mocha, Babel, Webpack and Rsbuild and am trying out new things on a new project here, and haven't set up testing yet.  I'm trying to feel out how well TanStack router will satisfy our needs.

When running Playwright tests in node via Vitest, would Vitest be bundling the code and running the bundles in node?  From experience with older tools I've been hesitant to run bundled code on the backend since debugging and source mapping has seemed to be more tricky with bundlers.  I already know how to register a transpiler module loader, but I'll also have to see how the debugging experience is with Vitest.  But if I could get away without any transpiling or bundling using the latest Node I would love to, because it would be that much faster and more straightforward to debug.

I'm also thinking, importing the router feels heavy when I just need to build some URLs, since that would load many of my app modules.  In other existing large projects we have loading an entire such router would be slow, at least without using dynamic imports everywhere so it doesn't load virtually the entire project.

**Reply by schiller-manuel** [reply]
i think vitest does not bundle but transpile on the fly, could be wrong though.

**Reply by jedwards1211** [reply]
@schiller-manuel what is the proper way to initialize the router in a Node environment (without jsdom or `document` or anything)?

When I just have `const router = createRouter({ routeTree })` in a Node environment, calling `buildLocation` errors out:

```
TypeError: Cannot read properties of undefined (reading 'pathname')
 ❯ Router.matchRoutesInternal ../../node_modules/.pnpm/@tanstack+router-core@1.134.12/node_modules/@tanstack/router-core/dist/esm/router.js:960:12
 ❯ RouterCore.matchRoutes ../../node_modules/.pnpm/@tanstack+router-core@1.134.12/node_modules/@tanstack/router-core/dist/esm/router.js:250:19
 ❯ build ../../node_modules/.pnpm/@tanstack+router-core@1.134.12/node_modules/@tanstack/router-core/dist/esm/router.js:279:48
 ❯ buildWithMatches ../../node_modules/.pnpm/@tanstack+router-core@1.134.12/node_modules/@tanstack/router-core/dist/esm/router.js:382:22
 ❯ RouterCore.buildLocation ../../node_modules/.pnpm/@tanstack+router-core@1.134.12/node_modules/@tanstack/router-core/dist/esm/router.js:423:14
```

I can't seem to see a way to initialize its `latestLocation` without having a History API instance...shouldn't we have an easy way to `buildLocation` that doesn't depend on any browser-only APIs?

**Reply by jedwards1211** [reply]
Okay I was able to get it to work with

```ts
import { describe, it, expect } from "vitest";
import { createRouter } from "@tanstack/react-router";
import { routeTree } from "../src/client/routeTree.gen.ts";
import { createMemoryHistory } from "@tanstack/history";

const router = createRouter({
  routeTree,
  history: createMemoryHistory({ initialEntries: ["/"], initialIndex: 0 }),
});

describe(`testRouter.buildLocation`, function () {
  it(`works`, function () {
    expect(
      router.buildLocation({
        to: "/users/$userId",
        params: { userId: 2 },
      }).pathname
    ).to.equal("/users/2");
  });
});

```

The docs were of little help to me here though -- how to do testing, especially end-to-end testing with browser automation, is too neglected in the docs, and it feels kind of like an afterthought in the API design as well.

I am being a grouch, and maybe my expectations are unfair, but FWIW I thought it would be way simpler to just build URLs for Playwright in a typechecked manner than this.

**Reply by schiller-manuel** [reply]
we are working on this, expect cool things to come in this area from @beaussan
